[Keyboard] add support for the kinT kinesis keyboard controller (kint36 variant)

stapelberg opened the pull request:
## Description

This PR adds support for the kinT kinesis keyboard controller, which is a new revision of the older stapelberg controller.

See https://github.com/kinx-project/kint for the repository, https://michael.stapelberg.ch/posts/2020-07-09-kint-kinesis-keyboard-controller/ for the introduction blog post and https://youtu.be/I0kwQbnhlfk for the introduction stream, if you’re curious about more details.

The commit to track a different version of ChibiOS-Contrib of course needs to be removed once https://github.com/qmk/ChibiOS-Contrib/pull/13 is in and before committing, but I wanted to get the review of the rest of this PR started already :)

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented:
This isn't just a keyboard, it's a core change. 

This should be broken out into 2 separate PRs, as per the guidelines.

drashna commented on keyboards/kinesis/kint36/config.h:
```diff
@@ -0,0 +1,71 @@
+ *                  ROW2COL = ROW = Anode (+), COL = Cathode (-, marked on diode)
+ *
+ */
+#define MATRIX_ROW_PINS  \
+    {                    \
+        LINE_PIN8,       \
+            LINE_PIN9,   \
+            LINE_PIN10,  \
+            LINE_PIN11,  \
+            LINE_PIN7,   \
+            LINE_PIN16,  \
+            LINE_PIN5,   \
+            LINE_PIN3,   \
+            LINE_PIN4,   \
+            LINE_PIN1,   \
+            LINE_PIN0,   \
+            LINE_PIN2,   \
+            LINE_PIN17,  \
+            LINE_PIN23,  \
+            LINE_PIN21,  \
+    }
+
+#define MATRIX_COL_PINS  \
+    {                    \
+        LINE_PIN18,      \
+            LINE_PIN14,  \
+            LINE_PIN15,  \
+            LINE_PIN20,  \
+            LINE_PIN22,  \
+            LINE_PIN19,  \
+            LINE_PIN6    \
+    }
```
Why not use the short notation, such as A5, etc?

stapelberg replied:
The short notation is not consistently available.

Take a look at https://www.pjrc.com/teensy/card9a_rev1.pdf, which is shipped with every Teensy 3.6 as a physical card.

It contains the numerical pin numbers for every pin, but only a few pins (mostly on the top side of the Teensy 3.6) are named with `A*`.

The situation is the same with the device symbol in the schematic: https://github.com/kinx-project/kint/blob/master/schematic-v2020-06-30.pdf

zvecr replied:
Thats not really how the mapping works within QMK, the `A*` stuff is Arduino notation. <https://www.pjrc.com/teensy/schematic.html> but minus the `PT` prefix should be used (from memory, sometimes you need to dig into the chips actual datasheet).

stapelberg replied:
I tried defining the short pin names in `quantum/config_common.h`, but defining e.g. `C1`, `C2`, `C5`, `C6` results in compilation errors in `lib/chibios-contrib/os/hal/ports/KINETIS/MK66F18/hal_lld.c` and others:

```
lib/chibios-contrib/os/hal/ports/KINETIS/MK66F18/hal_lld.c: In function 'MK66F18_clock_init':
./lib/chibios-contrib/os/hal/ports/KINETIS/LLD/GPIOv1/hal_pal_lld.h:154:3: error: expected identifier before '(' token
  154 |   ((ioline_t)((uint32_t)(port) | ((uint32_t)(pad)<<20)))
      |   ^
./lib/chibios-contrib/os/hal/boards/PJRC_TEENSY_3_6/board.h:296:29: note: in expansion of macro 'PAL_LINE'
  296 | #define LINE_PIN23          PAL_LINE(TEENSY_PIN23_IOPORT, TEENSY_PIN23)
      |                             ^~~~~~~~
quantum/config_common.h:186:20: note: in expansion of macro 'LINE_PIN23'
  186 | #        define C2 LINE_PIN23
      |                    ^~~~~~~~~~
lib/chibios-contrib/os/hal/ports/KINETIS/MK66F18/hal_lld.c:194:10: note: in expansion of macro 'C2'
  194 |     MCG->C2 = MCG_C2_LOCRE0 | MCG_C2_EREFS0 | MCG_C2_RANGE0_SET(2);
      |          ^~

```

Are you sure we need the short names and can’t use the `LINE_PIN*` names from https://github.com/qmk/ChibiOS-Contrib/blob/e3a3a24047717af33b098266e11c2e4e5102206a/os/hal/boards/PJRC_TEENSY_3_6/board.h#L273 ?

drashna commented on keyboards/kinesis/kint36/kint36.c:
```diff
@@ -0,0 +1,62 @@
+    matrix_init_user();
+}
+
+#define ledTurnOff palSetLine
+#define ledTurnOn palClearLine
```
please convert to GPIO control functions: 
https://docs.qmk.fm/#/internals_gpio_control

stapelberg replied:
Done

drashna commented on keyboards/kinesis/kint36/kint36.h:
```diff
@@ -0,0 +1,76 @@
+#define LAYOUT(                                   \
+    kC0, kD0, kE0, kC1, kD1, kE1, kC2, kD2, kE2,  \
+    k00, k10, k20, k30, k40, k50,                 \
+    k01, k11, k21, k31, k41, k51,                 \
+    k02, k12, k22, k32, k42, k52,                 \
+    k03, k13, k23, k33, k43, k53,                 \
+         k14, k24, k34, k54,                      \
+                             k56, k55,            \
+                                  k35,            \
+                        k36, k46, k25,            \
+                                                  \
+    kC3, kD3, kE3, kC4, kD4, kE4, kC5, kE5, kD5,  \
+                   k60, k70, k80, k90, kA0, kB0,  \
+                   k61, k71, k81, k91, kA1, kB1,  \
+                   k62, k72, k82, k92, kA2, kB2,  \
+                   k63, k73, k83, k93, kA3, kB3,  \
+                        k64, k84, k94, kA4,       \
+              k96, k85,                           \
+              k86,                                \
+              k66, k75, k65                       \
+) {                                              \
+    { k00,  k01,  k02,  k03,  ___,  ___,  ___ }, \
+    { k10,  k11,  k12,  k13,  k14,  ___,  ___ }, \
+    { k20,  k21,  k22,  k23,  k24,  k25,  ___ }, \
+    { k30,  k31,  k32,  k33,  k34,  k35,  k36 }, \
+    { k40,  k41,  k42,  k43,  ___,  ___,  k46 }, \
+    { k50,  k51,  k52,  k53,  k54,  k55,  k56 }, \
+    { k60,  k61,  k62,  k63,  k64,  k65,  k66 }, \
+    { k70,  k71,  k72,  k73,  ___,  k75,  ___ }, \
+    { k80,  k81,  k82,  k83,  k84,  k85,  k86 }, \
+    { k90,  k91,  k92,  k93,  k94,  ___,  k96 }, \
+    { kA0,  kA1,  kA2,  kA3,  kA4,  ___,  ___ }, \
+    { kB0,  kB1,  kB2,  kB3,  ___,  ___,  ___ }, \
+    { kC0,  kC1,  kC2,  kC3,  kC4,  kC5,  ___ }, \
+    { kD0,  kD1,  kD2,  kD3,  kD4,  kD5,  ___ }, \
+    { kE0,  kE1,  kE2,  kE3,  kE4,  kE5,  ___ }, \
+}
+
+/*  ---------------- LEFT HAND -----------------   ---------------- RIGHT HAND ---------------- */
+#define LAYOUT_pretty(                                                                           \
```
```suggestion
#define LAYOUT(                                   \
```

stapelberg replied:
I can’t seem to commit your suggestion? This is what I see:

![suggestion](https://user-images.githubusercontent.com/55506/91631922-954b3680-e9dd-11ea-8df6-b7824e50fab9.jpg)

Is there a mistake somewhere, or could you spell out what you’d like to see changed please?

noroadsleft replied:
Personally, I'm thinking this change shouldn't be made, because it maintains compatibility with existing keymaps that may exist for this keyboard. Perhaps a user who already has one of the older controllers may want or need to upgrade later.

stapelberg replied:
Sorry, I don’t follow. What change are you against, and what compatibility are you concerned with? The kint36 variant I’m introducing here isn’t compatible with the old stapelberg controllers to begin with, so I don’t think that plays a role here?

stapelberg commented:
> This isn't just a keyboard, it's a core change.
> 
> This should be broken out into 2 separate PRs, as per the guidelines.

Are you referring to commit https://github.com/qmk/qmk_firmware/pull/10171/commits/6743a384139edf2c4dfd6d82510865047ab6d97b, or are you saying that commit https://github.com/qmk/qmk_firmware/pull/10171/commits/e96e6b5f395a8f86ed4263567b18262a4fd3406a needs to be split up further?

If the former, please see https://github.com/qmk/ChibiOS-Contrib/pull/13 (it’s already split out).

If the latter, can you explain what exactly needs to be split out please?

tzarc commented:
At minimum, this will need to wait for the ChibiOS 20.x upgrade to come through.
The ChibiOS-Contrib 20.3.x branch requires ChibiOS 20.3.x -- we're currently using 19.1.3.

I've already got a branch in progress to perform the ChibiOS upgrades across the platform -- the holdup will be the fact that every single ARM keyboard breaks unless they're upgraded at the same time. This is mostly in progress -- F303 and F072 are largely complete: [conf branch here](https://github.com/tzarc/qmk_firmware/pull/new/chibios-conf-migration), [chibios upgrade branch here](https://github.com/tzarc/qmk_firmware/commits/generated-chibios-master-upgrade), [chibios conf upgrader branch here](https://github.com/tzarc/qmk_firmware/commits/chibiconf-migrator). F1, F0, and F042 are still outstanding.

Issue to track: #8976.

tzarc commented:
> If the latter, can you explain what exactly needs to be split out please?

Basically anything outside `keyboards/` or `users/` -- in this case it's the makefile changes and the modifications to `config_common.h` and `bootloader.c`.

For the record, I don't see any issue with the nature of those changes specifically -- the issue will be waiting for the ChibiOS upgrade to land on `develop`.

stapelberg commented:
This is still relevant, but waiting for https://github.com/qmk/ChibiOS-Contrib to be updated to a more recent ChibiOS-Contrib.

tzarc commented:
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

drashna commented:
There are a bunch of merge conflicts here. Probably needs a rebase.

stapelberg commented:
> There are a bunch of merge conflicts here. Probably needs a rebase.

Thanks for the note.

I noticed that the commits I had been waiting for are now available, so I rebased my changes onto the current `develop` branch.

This should now be ready for review. Please take another look :)

fauxpark commented on data/schemas/keyboard.jsonschema:
```diff
@@ -25,7 +25,7 @@
         "processor": {
             "type": "string",
-            "enum": ["cortex-m0", "cortex-m0plus", "cortex-m3", "cortex-m4", "MKL26Z64", "MK20DX128", "MK20DX256", "STM32F042", "STM32F072", "STM32F103", "STM32F303", "STM32F401", "STM32F411", "STM32G431", "STM32G474", "atmega16u2", "atmega32u2", "atmega16u4", "atmega32u4", "at90usb162", "at90usb646", "at90usb647", "at90usb1286", "at90usb1287", "atmega32a", "atmega328p", "atmega328", "attiny85",  ...
+            "enum": ["cortex-m0", "cortex-m0plus", "cortex-m3", "cortex-m4", "MKL26Z64", "MK20DX128", "MK20DX256", "MK66F18", "STM32F042", "STM32F072", "STM32F103", "STM32F303", "STM32F401", "STM32F411", "STM32G431", "STM32G474", "atmega16u2", "atmega32u2", "atmega16u4", "atmega32u4", "at90usb162", "at90usb646", "at90usb647", "at90usb1286", "at90usb1287", "atmega32a", "atmega328p", "atmega328", " ...
```
The core stuff should be done in a separate PR.

stapelberg replied:
Done, sent https://github.com/qmk/qmk_firmware/pull/12258

fauxpark commented on keyboards/kinesis/kint36/chconf.h:
```diff
@@ -0,0 +1,755 @@
+/*
```
Please run `qmk chibios-confmigrate` on these files.

stapelberg replied:
Done

fauxpark commented on tmk_core/common/chibios/pin_defs.h:
```diff
@@ -53,6 +53,9 @@
 #        define D5 PAL_LINE(GPIOC, 13)
 #        define B0 PAL_LINE(GPIOC, 14)
 #    endif
+#elif defined(KEYBOARD_kinesis_kint36)
+// The compatibility defines do not work with the KINETIS header files,
+// so do not declare anything.
```
```suggestion
```

stapelberg replied:
See the other comment on the same line below regarding define availability

tzarc commented on keyboards/kinesis/kint36/kint36.c:
```diff
@@ -0,0 +1,62 @@
+#include "kint36.h"
+#include "chprintf.h"
```
```suggestion
```

stapelberg replied:
Done

tzarc reviewed (changes requested):
We'll also need GPLv2+ compatible license headers on each source file.

stapelberg commented on tmk_core/common/chibios/pin_defs.h:
```diff
@@ -53,6 +53,9 @@
 #        define D5 PAL_LINE(GPIOC, 13)
 #        define B0 PAL_LINE(GPIOC, 14)
 #    endif
+#elif defined(KEYBOARD_kinesis_kint36)
```
It is definitely required (build fails loudly without), but unfortunately MK66F18 is not defined here. It *is* defined in bootloader.c, though (add an `#error` directive to see quickly).

In previous revisions of this PR, I set `OPT_DEFS += -DQMK_KINETIS` in `build_keyboard.mk`, but that was refactored and I think removed? Where should I be setting extra `-D` flags, if any?

stapelberg commented on keyboards/kinesis/kint36/chconf.h:
```diff
@@ -0,0 +1,755 @@
+/*
```
Done

stapelberg commented on keyboards/kinesis/kint36/halconf.h:
```diff
@@ -0,0 +1,533 @@
+// copied from infinity60/halconf.h
```
Done

stapelberg commented:
> We'll also need GPLv2+ compatible license headers on each source file.

Done

stapelberg commented:
Now rebased and all comments addressed, I think.

Please take another look :)

tzarc commented on keyboards/kinesis/kint36/chconf.h:
```diff
@@ -0,0 +1,129 @@
+/* Copyright 2020 QMK
```
Are the majority of these necessary? Seems like the original config must've been with everything enabled?
Building locally after deleting this file ends up producing a binary -- might want to check if it can just be removed.

stapelberg replied:
chconf.h now minified. The remaining settings need to be present, otherwise the generated binary doesn’t work on the Teensy 3.6.

tzarc commented on keyboards/kinesis/kint36/halconf.h:
```diff
@@ -0,0 +1,27 @@
+
+#pragma once
+
+#define HAL_USE_I2C TRUE
```
Does your build use I2C? If not, this file should be able to be safely removed.

stapelberg replied:
halconf.h now deleted

tzarc reviewed (changes requested):
Just a couple of queries.

fauxpark commented on keyboards/kinesis/kint36/config.h:
```diff
@@ -0,0 +1,87 @@
+ */
+#define MATRIX_ROW_PINS  \
+    {                    \
+        LINE_PIN8,       \
```
These can be changed to use the "normal" pin names now.

stapelberg replied:
Repeating my other earlier comment:

I tried defining the short pin names in `quantum/config_common.h`, but defining e.g. `C1`, `C2`, `C5`, `C6` results in compilation errors in `lib/chibios-contrib/os/hal/ports/KINETIS/MK66F18/hal_lld.c` and others:

```
lib/chibios-contrib/os/hal/ports/KINETIS/MK66F18/hal_lld.c: In function 'MK66F18_clock_init':
./lib/chibios-contrib/os/hal/ports/KINETIS/LLD/GPIOv1/hal_pal_lld.h:154:3: error: expected identifier before '(' token
  154 |   ((ioline_t)((uint32_t)(port) | ((uint32_t)(pad)<<20)))
      |   ^
./lib/chibios-contrib/os/hal/boards/PJRC_TEENSY_3_6/board.h:296:29: note: in expansion of macro 'PAL_LINE'
  296 | #define LINE_PIN23          PAL_LINE(TEENSY_PIN23_IOPORT, TEENSY_PIN23)
      |                             ^~~~~~~~
quantum/config_common.h:186:20: note: in expansion of macro 'LINE_PIN23'
  186 | #        define C2 LINE_PIN23
      |                    ^~~~~~~~~~
lib/chibios-contrib/os/hal/ports/KINETIS/MK66F18/hal_lld.c:194:10: note: in expansion of macro 'C2'
  194 |     MCG->C2 = MCG_C2_LOCRE0 | MCG_C2_EREFS0 | MCG_C2_RANGE0_SET(2);
      |          ^~

```

Are you sure we need the short names and can’t use the `LINE_PIN*` names from https://github.com/qmk/ChibiOS-Contrib/blob/e3a3a24047717af33b098266e11c2e4e5102206a/os/hal/boards/PJRC_TEENSY_3_6/board.h#L273 ?

---

Are those the normal pin names you mean?

If no, can you please spell out specifically what your suggestion is?

fauxpark replied:
It works if you don't include config_common.h, as I said earlier.

stapelberg replied:
Indeed, it seems to compile now.

I’m still confused, though: what actually are the short names? (Arduino pinout?) How do I know which short name corresponds to which Teensy pin?

From the Teensy perspective, https://www.pjrc.com/teensy/card11a_rev3_web.pdf documents the pinout. You can see that all pins (aside from power pins) are just numbered.

fauxpark replied:
For that you'd have to look at the datasheet, they are prefixed with `PT`: https://www.pjrc.com/teensy/K66P144M180SF5V2.pdf#page=72

The silkscreen numbering is the Arduino pinout, so it's pretty much useless for QMK and doesn't really correspond to anything in the datasheet either.

stapelberg replied:
Got it now, thank you! Done.

fauxpark commented on keyboards/kinesis/kint36/config.h:
```diff
@@ -0,0 +1,87 @@
+#undef DEVICE_VER
+#define DEVICE_VER 0x0001
+#undef MANUFACTURER
+#define MANUFACTURER "https://github.com/stapelberg"
+#undef PRODUCT
+#define PRODUCT "https://github.com/kinx-project/kint"
```
These shouldn't really be URLs...

stapelberg replied:
Huh, why not? I find URLs to be the most elegant thing to put into the product field for self-built devices. That way, it’s very clear where to find more information/sources :)

fauxpark replied:
It's meant to be the product *name*, not the URL. It would just look weird.
![image](https://user-images.githubusercontent.com/4781841/111449230-4ee3dc00-8763-11eb-8ec7-151f221d57c1.png)

stapelberg replied:
Done.

stapelberg commented:
Should I split the change to tmk_core/common/chibios/pin_defs.h into its own PR again?

stapelberg commented:
> Should I split the change to tmk_core/common/chibios/pin_defs.h into its own PR again?

I’ll assume yes. The core change is now split out into https://github.com/qmk/qmk_firmware/pull/12419

stapelberg commented:
This should be ready to merge now that https://github.com/qmk/qmk_firmware/commit/e95a4f4f52f7792927c480e76dba11c5a8cbd07e has landed. I have rebased my change accordingly.

Thanks!

fauxpark commented on keyboards/kinesis/kint36/config.h:
```diff
@@ -0,0 +1,87 @@
+ *                  ROW2COL = ROW = Anode (+), COL = Cathode (-, marked on diode)
+ *
+ */
+#define MATRIX_ROW_PINS  \
+    {                    \
+        D3,       \
+            C3,   \
+            C4,   \
+            C6,   \
+            D2,   \
+            B0,   \
+            D7,   \
+            A12,  \
+            A13,  \
+            B17,  \
+            B16,  \
+            D0,   \
+            B1,   \
+            C2,   \
+            D6,   \
+    }
+
+#define MATRIX_COL_PINS  \
+    {                    \
+        B3,      \
+            D1,  \
+            C0,  \
+            D5,  \
+            C1,  \
+            B2,  \
+            D4   \
+    }
```
```suggestion
#define MATRIX_ROW_PINS { D3, C3, C4, C6, D2, B0, D7, A12, A13, B17, B16, D0, B1, C2, D6 }
#define MATRIX_COL_PINS { B3, D1, C0, D5, C1, B2, D4 }
```

stapelberg replied:
Done!

fauxpark commented on keyboards/kinesis/kint36/kint36.c:
```diff
@@ -0,0 +1,77 @@
+#define ledTurnOn writePinLow
+
+#define LED_KEYPAD E26
+#define LED_SCROLL_LOCK A5
+#define LED_NUM_LOCK A14
+#define LED_CAPS_LOCK C7
```
https://docs.qmk.fm/#/feature_led_indicators

stapelberg replied:
Done, thanks for the hint!

stapelberg commented:
Thanks for the review, please take another look :)
